File: Server/src/main/content/global/skill/production/crafting/item/HardCraftPulse.java
package content.global.skill.production.crafting.item;

import content.global.skill.production.crafting.data.LeatherData;
import core.game.node.entity.player.Player;
import core.game.node.entity.skill.SkillPulse;
import core.game.node.entity.skill.Skills;
import core.game.node.item.Item;
import core.game.world.update.flag.context.Animation;

public final class HardCraftPulse extends SkillPulse<Item> {

    private static final Animation ANIMATION = Animation.create(1249);
    private int amount;
    private int ticks;

    public HardCraftPulse(Player player, Item node, int amount) {
        super(player, node);
        this.amount = amount;
    }

    @Override
    public boolean checkRequirements() {
        if (player.getSkills().getLevel(Skills.CRAFTING) < 28) {
            player.getDialogueInterpreter().sendDialogue("You need a crafting level of " + 28 + " to make a hardleather body.");
            return false;
        }
        if (!player.getInventory().contains(LeatherData.NEEDLE, 1)) {
            return false;
        }
        if (!player.getInventory().contains(LeatherData.HARD_LEATHER, 1)) {
            return false;
        }
        if (!player.getInventory().containsItem(LeatherData.THREAD)) {
            player.getDialogueInterpreter().sendDialogue("You need thread to make this.");
            return false;
        }
        player.getInterfaceManager().close();
        return true;
    }

    @Override
    public void animate() {
        if (ticks % 5 == 0) {
            player.animate(ANIMATION);
        }
    }

    @Override
    public boolean reward() {
        if (++ticks % 5 != 0) {
            return false;
        }
        if (player.getInventory().remove(new Item(LeatherData.HARD_LEATHER))) {
            Item item = new Item(1131);
            player.getInventory().add(item);
            player.getSkills().addExperience(Skills.CRAFTING, 35, true);
            LeatherData.decayThread(player);
            if (LeatherData.isLastThread(player)) {
                LeatherData.removeThread(player);
            }
        }
        amount--;
        return amount < 1;
    }

}

File: Server/src/main/content/global/skill/production/crafting/item/HardCraftPulse.kt
package content.global.skill.production.crafting.item

import content.global.skill.production.crafting.data.LeatherData
import content.global.skill.production.crafting.data.LeatherData.decayThread
import content.global.skill.production.crafting.data.LeatherData.isLastThread
import content.global.skill.production.crafting.data.LeatherData.removeThread
import core.api.*
import core.api.consts.Animations
import core.api.consts.Items
import core.game.node.entity.player.Player
import core.game.node.entity.skill.SkillPulse
import core.game.node.entity.skill.Skills
import core.game.node.item.Item

/**
 * Hard craft pulse
 *
 * @param amount The amount of items to be crafted.
 * @constructor Represents a new instance of HardCraftPulse.
 *
 * @param player The player who is performing the crafting action.
 * @param node The item node that represents the item being crafted.
 */
class HardCraftPulse(player: Player?, node: Item?, var amount: Int) : SkillPulse<Item?>(player, node) {

    var ticks = 0

    override fun checkRequirements(): Boolean {
        if (getStatLevel(player, Skills.CRAFTING) < 28) {
            sendDialogue(player, "You need a crafting level of " + 28 + " to make a hardleather body.")
            return false
        }
        if (!inInventory(player, LeatherData.NEEDLE, 1)) {
            return false
        }
        if (!inInventory(player,LeatherData.HARD_LEATHER, 1)) {
            return false
        }
        if (!inInventory(player, LeatherData.THREAD.id)) {
            sendDialogue(player, "You need thread to make this.")
            return false
        }
        closeInterface(player)
        return true
    }

    override fun animate() {
        if (ticks % 5 == 0) {
            animate(player, ANIMATION)
        }
    }

    override fun reward(): Boolean {
        if (++ticks % 5 != 0) {
            return false
        }
        if (removeItem(player, Item(LeatherData.HARD_LEATHER))) {
            val item = Item(Items.HARDLEATHER_BODY_1131)
            addItem(player, item.id)
            rewardXP(player, Skills.CRAFTING, 35.0)
            decayThread(player)
            if (isLastThread(player)) {
                removeThread(player)
            }
        }
        amount--
        return amount < 1
    }

    companion object {
        private const val ANIMATION = Animations.CRAFT_LEATHER_1249
    }
}
